### ISSUE TITLE
Error Rate Not Reflective of Output Sequences

### ISSUE THREAD

**Issue by elithrar** [BrainJS/brain.js#375]
![???](https://media.giphy.com/media/53lN577WPCFGM/giphy.gif)

## *What* is wrong?

The error rate approaches ~0.011 [as set] but the network struggles to learn the correct sequences. 

* I've experimented with different layer configurations, although this isn't a hugely diverse corpus and I'd expect ~30 neurons for the # of unique inputs (char-based RNN; so low) to be acceptable. This doesn't seem to improve with 128, 256 or even 1024 (which takes too long to train in Node.js) neurons.
* I clean up the input to remove empty newlines and [at this point] the name of each speaker/character in the play.

## *Where* does it happen?

After ~50k iterations of training against [a subset of the Shakespeare](https://gist.github.com/elithrar/fbcca240e1bdeffd9629b94dc002a491) corpus

```
function trainRNN(inputs, options) {
    // 1. Create our neural network
    // TODO(matt): Provide config options for network shape.
    const net = new brain.recurrent.GRU({
        hiddenLayers: [30, 30],
    })

    let iters = options.iterations
    let logger = (log) => {
        console.log(`\t⚪ ${log}\n\toutput: ${net.run(options.seed)}`)
    }

    // 3. Train it
    console.log(`⏳ starting training (iterations: ${iters})`)
    let summary = net.train(inputs, {
        iterations: iters,
        log: logger,
        logPeriod: iters / 50,
        errorThresh: 0.011,
    })

    console.log(summary)

    // Serialize our model & weights to JSON
    return net.toJSON()
}
```

Training config & data cleanup:

```
➜  yarn run train --epochs=30000 --dataset data/shakespeare.txt --seed "speak" --model-dir models/shakespeare.model.json
yarn run v1.13.0 
$ node src/train-brain.js --epochs=30000 --dataset data/shakespeare.txt --seed speak --model-dir models/shakespeare.model.json 
[ 'first citizen',
  'before we proceed any further, hear me speak',
  'speak, speak',
  'first citizen',
  'you are all resolved rather to die than to famish?',
  'resolved resolved',
  'first citizen',
  'first, you know caius marcius is chief enemy to the people', 
  "we know't, we know't",
  'first citizen',
  "let us kill him, and we'll have corn at our own price",      
  "is't a verdict?",
  "no more talking on't; let it be done away, away!",
  'second citizen',
  'one word, good citizens',
  'first citizen',
  'we are accounted poor citizens, the patricians good',        
  'what authority surfeits on would relieve us if they',        
  'would yield us but the superfluity, while it were',
  'wholesome, we might guess they relieved us humanely;' ]      
Sentences: 349
⏳ starting training (iterations: 30000) 
```

## *How* do we replicate the issue?

Train a GRU or LSTM RNN against https://gist.github.com/elithrar/fbcca240e1bdeffd9629b94dc002a491

## *How* important is this (1-5)?

3 - it's challenging to use Brain.js for anything but simple examples. I'd like for it to be able to compete with TensorFlow.js for web-facing / JS-requiring use-cases!

## Other Comments

Trawling through the code-base, the default `DataFormatter`[1] builds a char-based model given string input, which is what I'd expect. The challenge is that - even after 50k iterations - the error rate tends to bounce between Infinity and close to the target rate (e.g. 0.011), but the predicted text is nonsensical given the input. It's worth noting that it never reaches the target error rate; it seems like the network has a hard time with it's gradients and is potentially being caught in some local minima?

The examples - "doe, a deer, a female deer" and the children's book - are both very simple and only seem to handle a limited number of input sequences.

Compared to a similar network configuration (although with an Embedding layer) in TensorFlow a 128-neuron LSTM/GRU - the Brain.js based network never seems to learn sequences, despite an error rate that should start to show promise.

[1]: https://github.com/BrainJS/brain.js/blob/59c590e764810892096dd7298ddb83bbe06a9616/src/utilities/data-formatter.js

**Comment by elithrar** [comment]
Note: I'm training a 128-neuron (single layer) GRU as an example - with error rates & predictions every ~500 iters - to give a solid example. My tmux history doesn't have prior runs.
